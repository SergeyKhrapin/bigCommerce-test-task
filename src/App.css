.App {
  text-align: center;
}

.App.cartPopup-opened {
    overflow: hidden;
    height: 100vh;
}

.App-logo {
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
}

.App-intro {
  font-size: large;
}
