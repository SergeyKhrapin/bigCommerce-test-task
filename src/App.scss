.App {
    text-align: center;
    &-header {
        display: flex;
        align-items: center;
        &-logo {
            height: 80px;
            object-fit: cover;
        }
        &-nav {
            display: flex;
            height: 100%;
            &--list {
                display: flex;
                align-items: center;
                li {
                    margin: 0 20px;
                }
            }
        }
    }
}

.App.cartPopup-opened {
    overflow: hidden;
    height: 100vh;
}
