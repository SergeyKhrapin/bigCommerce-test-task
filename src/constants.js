export const PRUDUCTS_UNAVAILABLE_MESSAGE     = 'The products are unavailable. Sorry for inconvenience.';
export const VIEW_DETAILS                     = 'view details';
export const ADD_TO_CART                      = 'add to cart';
export const SERVER_URL                       = 'products.json';
export const INCREASE                         = 'INCREASE';
export const DECREASE                         = 'DECREASE';
export const ONE_PRODUCT_MAX_QUANTITY_IN_CART = 5;
export const TOTAL_MAX_QUANTITY_IN_CART       = 20;
export const CURRENCY                         = '$';

// Cart
export const CART_HEADING                     = 'Shopping Cart';
export const CART_EMPTY                       = 'Your cart is empty.'
export const PRODUCT                          = 'Product';
export const QUANTITY                         = 'Quantity';
export const TOTAL                            = 'Total';
export const SUBTOTAL                         = 'Subtotal';
export const ACTION                           = 'Action';
export const COLOR                            = 'Color';
export const CART_OVERVIEW                    = 'Cart overview';
export const CONTINUE_SHOPPING                = 'Continue shopping';
export const CHECKOUT                         = 'Checkout';
