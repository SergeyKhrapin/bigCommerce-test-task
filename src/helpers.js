export const getImageSrc = src => `/media/${src}`;

export const getFormattedPrice = price => Number(price).toFixed(2);

export const getProductDetails = (products, id) => products.find(product => parseInt(id) === parseInt(product.id));

export const getPrettyTitle = title => title.replaceAll(/\s/g, '-');
