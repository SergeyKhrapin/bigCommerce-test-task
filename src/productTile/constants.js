export const VIEW_DETAILS = 'view details';
export const ADD_TO_CART = 'add to cart';
export const CURRENCY = '$';
