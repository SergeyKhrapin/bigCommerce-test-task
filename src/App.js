import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';
import logo from './logo.svg';
import Category from "./category/Category";
import Cart from "./cart/Cart";
import Product from "./product/Product";
import CartPopup from './cartPopup/CartPopup';
import './App.css';

class App extends Component {
    constructor() {
        super();
    }

    render() {
        return (
            <div className={`App${this.props.openPopup ? ' cartPopup-opened' : ''}`}>
                {/* <header className="App-header">
                    <img src={logo} className="App-logo" alt="logo" />
                    <h1 className="App-title">Storefront Assignment</h1>
                </header>
                <p className="App-intro">
                    To get started, delete this header and introduction, and begin building your app in the provided components.
        </p>
                <p className="App-intro">
                    We've setup the bare minimum you need to get started, but feel free to add as many components as you see fit.
        </p> */}

                <CartPopup />

                <Route exact path="/" component={Category} />
                <Route path="/cart" component={Cart} />
                <Route path="/product/:id/:title" component={Product} />
            </div>
        );
    }
}

const mapStateToProps = state => {
    return {
        openPopup: state.cart.openPopup
    };
};

export default connect(mapStateToProps)(App);
