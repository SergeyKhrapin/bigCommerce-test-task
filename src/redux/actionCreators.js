import { SERVER_URL } from "../constants";
import { FETCH_PRODUCTS, ADD_PRODUCT_TO_CART, DECREASE_PRODUCT_QUANTITY_IN_CART, REMOVE_PRODUCT_FROM_CART } from "./actionTypes";

export const fetchProducts = () => {
    return async dispatch => {
        const response     = await fetch(SERVER_URL);
        const responseJSON = await response.json();

        dispatch({
            type: FETCH_PRODUCTS,
            payload: responseJSON
        })
    }
}

export const addProductToCart = (product, quantity = 1) => {
    // Default value quantity = 1 is used when a product is added to cart from PLP or its quantity increased from Cart page
    return {
        type: ADD_PRODUCT_TO_CART,
        payload: { product, quantity }
    }
}

export const decreaseProductQuantityInCart = (product, quantity = 1) => {
    return {
        type: DECREASE_PRODUCT_QUANTITY_IN_CART,
        payload: { product, quantity }
    }
}

export const removeProductFromCart = (product, quantity) => {
    return {
        type: REMOVE_PRODUCT_FROM_CART,
        payload: { product, quantity }
    }
}
