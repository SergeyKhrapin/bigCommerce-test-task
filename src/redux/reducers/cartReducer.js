import { ADD_PRODUCT_TO_CART, REMOVE_PRODUCT_FROM_CART } from "../actionTypes";
import { ONE_PRODUCT_MAX_QUANTITY_IN_CART } from '../../constants';
import { showWarning } from "../../helpers";

const initialState = {
    products: {},
    quantity: 0,
    total: 0
};

export const cartReducer = (state = initialState, action) => {
    switch (action.type) {
        case ADD_PRODUCT_TO_CART:
            let { products } = state;
            let { payload: {product, quantity} } = action;

            for (let productID in products) {
                // If the product is already in the cart - increase its quantity
                if (productID == product.id) {
                    const thisProductQuantityInCart = products[productID].quantity;
                    if (thisProductQuantityInCart + quantity <= ONE_PRODUCT_MAX_QUANTITY_IN_CART) {
                        quantity += thisProductQuantityInCart;
                    } else {
                        // showWarning`Max limit per product in the cart - ${ONE_PRODUCT_MAX_QUANTITY_IN_CART} items. Please decrease the quantity.`;
                        alert(`Max limit per product in the cart - ${ONE_PRODUCT_MAX_QUANTITY_IN_CART} items. Please decrease the quantity.`);
                        return state;
                    }
                    break;
                }
            }

            return {
                ...state,
                products: {
                    ...state.products,
                    [product.id]: {product, quantity}
                }
            };
        case REMOVE_PRODUCT_FROM_CART:
            // do something
        default:
            return state;
    }
}
