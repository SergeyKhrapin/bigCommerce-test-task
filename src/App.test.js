import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import ReactDOM from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import App from './App';
import * as helpers from './helpers';
import * as actionCreators from './redux/actionCreators';
import * as actionTypes from './redux/actionTypes';
import { cartReducer } from './redux/reducers/cartReducer';
import * as constant from './constants';

const stateMock = {
    products: {
        1000001: {
            product: { id: 1000001, price: 50, title: "Heme" },
            quantity: 2
        },
        1000003: {
            product: { id: 1000003, price: 40, title: 'MASHIKO' },
            quantity: 3
        }
    },
    totalQuantity: 5,
    totalPrice: 220,
    alert: null,
    openPopup: null
};

const actionMock = {
    type: 'SOME_TYPE',
    payload: {
        product: { id: 1000003, price: 40, title: 'MASHIKO' },
        quantity: 3
    }
};

describe('Helper: ', () => {
    it('getFormattedPrice should transform number 40 to $40.00', () => {
        expect(helpers.getFormattedPrice(40)).toBe('$40.00');
    });

    it('getPrettyTitleURL should replace spaces with dashes', () => {
        expect(helpers.getPrettyTitleURL('Blue Stripe Stoneware Plate')).toBe('Blue-Stripe-Stoneware-Plate');
    });

    it('getProductDetails should return a product or empty object', () => {
        expect(helpers.getProductDetails([])).toBeDefined();
    });
});

describe('Action creator:', () => {
    const product = { id: 1000003, price: 40, title: 'MASHIKO' };
    const action = {
        type: actionTypes.DECREASE_PRODUCT_QUANTITY_IN_CART,
        payload: { product, quantity: 1 }
    }

    it('should return an action object', () => {
        expect(actionCreators.decreaseProductQuantityInCart(product)).toStrictEqual(action);
    });

    it('should return an object with type field', () => {
        expect(actionCreators.decreaseProductQuantityInCart()).toHaveProperty('type');
        expect(actionCreators.removeProductFromCart()).toHaveProperty('type');
        expect(actionCreators.hideAlert()).toHaveProperty('type');
        expect(actionCreators.openCartPopup()).toHaveProperty('type');
        expect(actionCreators.closeCartPopup()).toHaveProperty('type');
    });
});

describe('Reducer:', () => {
    it('should always return a state object', () => {
        const initialStateMock = {
            products: {},
            totalQuantity: 0,
            totalPrice: 0,
            alert: null,
            openPopup: null
        };
        const actionMock = {type: 'SOME_NON_EXISTENT_TYPE'};
        expect(cartReducer(initialStateMock, actionMock)).toEqual(initialStateMock);
    });
});

describe('Add product to the cart:', () => {
    let state, action;

    beforeEach(() => {
        state = {
            products: {},
            totalQuantity: 0,
            totalPrice: 0,
            alert: null,
            openPopup: null
        };
        action = {
            type: actionTypes.ADD_PRODUCT_TO_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 0
            }
        };
    });

    it('if total products quantity in cart will exceed max limit - product is not added and alert is displayed', () => {
        state.totalQuantity = 19;
        action.payload.quantity = 2;
        const updatedState = cartReducer(state, action);
        expect(updatedState.totalQuantity).toBeLessThanOrEqual(constant.TOTAL_MAX_QUANTITY_IN_CART);
        expect(updatedState.alert).toBe(constant.TOTAL_MAX_QUANTITY_ALERT);
    });

    it('if product quantity in cart will exceed max limit per one product - product is not added and alert is displayed', () => {
        action.payload.quantity = 6;
        const updatedState = cartReducer(state, action);
        expect(updatedState.products['1000003']).not.toBeDefined();
        expect(updatedState.alert).toBe(constant.ONE_PRODUCT_MAX_QUANTITY_ALERT);
    });
});

describe('Remove product from the cart should lead to:', () => {
    let state, action, updatedState;

    beforeEach(() => {
        state = {
            products: {
                1000001: {
                    product: { id: 1000001, price: 50, title: "Heme" },
                    quantity: 2
                },
                1000003: {
                    product: { id: 1000003, price: 40, title: 'MASHIKO' },
                    quantity: 3
                }
            },
            totalQuantity: 5,
            totalPrice: 220,
            alert: null,
            openPopup: null
        };

        action = {
            type: actionTypes.REMOVE_PRODUCT_FROM_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 3
            }
        };

        updatedState = cartReducer(state, action);
    });

    it('remove product id field from cart products object', () => {
        expect(updatedState.products).not.toHaveProperty([action.payload.product.id]);
    });

    it('reduce total products quantity by quantity of removed product', () => {
        expect(updatedState.totalQuantity).toEqual(state.totalQuantity - action.payload.quantity);
    });

    it('reduce total price by total price of removed product', () => {
        expect(updatedState.totalPrice).toEqual(state.totalPrice - action.payload.product.price * action.payload.quantity);
    });
});

describe('Decrease a product quantity in the cart should lead to:', () => {
    let state, action, updatedState;

    beforeEach(() => {
        state = {
            products: {
                1000001: {
                    product: { id: 1000001, price: 50, title: "Heme" },
                    quantity: 2
                },
                1000003: {
                    product: { id: 1000003, price: 40, title: 'MASHIKO' },
                    quantity: 3
                }
            },
            totalQuantity: 5,
            totalPrice: 220,
            alert: null,
            openPopup: null
        };

        action = {
            type: actionTypes.DECREASE_PRODUCT_QUANTITY_IN_CART,
            payload: {
                product: { id: 1000003, price: 40, title: 'MASHIKO' },
                quantity: 3
            }
        };

        updatedState = cartReducer(state, action);
    });

    it('decrease a product quantity by 1', () => {
        expect(updatedState.products['1000003'].quantity).toEqual(2);
    });

    it('decrease total products quantity by 1', () => {
        expect(updatedState.totalQuantity).toEqual(4);
    });

    it('decrease total price by product price', () => {
        expect(updatedState.totalPrice).toEqual(220 - 40);
    });
});

describe('App component renders:', () => {
    it('correctly without products', () => {
        const mockStore = configureStore();
        const store = mockStore({
            allProducts: [],
            cart: stateMock
        });

        const component = renderer.create(
            <Provider store={store}>
                <BrowserRouter>
                    <App />
                </BrowserRouter>
            </Provider>
        );

        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});

describe('Fetch products:', () => {
    const mockProducts = [
        { id: 1000001, price: 40, title: 'MASHIKO' },
        { id: 1000002, price: 60, title: 'HEME' },
    ];

    const action = {
        type: actionTypes.FETCH_PRODUCTS,
        payload: { allProducts: mockProducts }
    };

    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    it('should return an action object with products', async () => {
        window.fetch = jest.fn().mockImplementation(() => ({
            status: 200,
            json: () => new Promise((resolve, reject) => {
                resolve({
                    allProducts: mockProducts
                })
            })
        }))

        const store = mockStore({});

        await store.dispatch(actionCreators.fetchProducts());
        const actions = store.getActions();

        expect(actions[0]).toEqual(action);
    });
});

describe('Add product to cart:', () => {
    const initialStateMock = {
        products: {},
        totalQuantity: 0,
        totalPrice: 0,
        alert: null,
        openPopup: null
    };

    const mockProduct = { id: '1000001', price: 40, title: 'MASHIKO' };

    const action = {
        type: actionTypes.ADD_PRODUCT_TO_CART,
        payload: {
            product: mockProduct,
            quantity: 1
        }
    };

    const middlewares = [thunk];
    const mockStore = configureStore(middlewares);

    it('should return an action object with that product', async () => {
        const store = mockStore({
            allProducts: [],
            cart: initialStateMock
        });

        await store.dispatch(actionCreators.addProductToCart(mockProduct, 1));
        const actions = store.getActions();
        console.log('actions', actions);
        const state = store.getState();
        console.log('state', state);

        expect(actions[0]).toEqual(action);
    });
});

