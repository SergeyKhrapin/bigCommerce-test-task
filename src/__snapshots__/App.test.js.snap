// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders: correctly without products 1`] = `
<div
  className="App"
  data-testid="app-container"
>
  <header
    className="App-header"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-2"
        >
          <img
            alt="logo"
            className="App-header-logo"
            src="logo.png"
          />
        </div>
        <div
          className="col-8"
        >
          <nav
            className="App-header-nav"
          >
            <ul
              className="App-header-nav--list"
            >
              <li>
                <a
                  href="/"
                  onClick={[Function]}
                >
                  HOME
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  SHOP
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  JOURNAL
                </a>
              </li>
              <li>
                <a
                  href="#"
                >
                  MORE
                </a>
              </li>
            </ul>
          </nav>
        </div>
        <div
          className="col-2"
        >
          <div
            className="CartPopup-open"
          >
            <span
              aria-expanded="false"
              className="CartPopup-open--link"
              data-testid="cart-popup-open-link"
              onClick={[Function]}
            >
              My cart (5)
            </span>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    className="Category"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-12"
        >
          <div
            className="Category-banner"
          >
            <img
              alt="Home page banner"
              className="Category-banner--image"
              src="/media/plates-header.jpg"
            />
            <div
              className="Category-banner--content"
            >
              <h1>
                Plates
              </h1>
              <p>
                Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsum eveniet rem ad culpa quidem incidunt laborum accusamus eius earum animi aliquid corporis inventore ex, dignissimos consequuntur expedita, vitae distinctio odit.
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        className="row mt-5"
      >
        <h1>
          The products are unavailable. Sorry for inconvenience.
        </h1>
      </div>
    </div>
  </div>
</div>
`;
